Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift

```diff
@@ -0,0 +1,1277 @@
+      acceptableRequestEncoding: acceptableRequestEncoding,
+      userProvidedHeaders: nil,
+      normalizeUserProvidedHeaders: self.normalizeHeaders
+    )
+
+    return .init(
+      state: .requestClosedResponseClosed,
+      action: .write(.headers(.init(headers: trailers, endStream: true)), nil, flush: true)
+    )
+  }
+
+  /// Makes a 'GRPCStatus' and response trailers suitable for returning to the client when the
+  /// request message encoding is not supported.
+  ///
+  /// - Parameters:
+  ///   - encoding: The unsupported request message encoding sent by the client.
+  ///   - acceptable: The list if acceptable request message encoding the client may use.
+  /// - Returns: The status and trailers to return to the client.
+  private func makeStatusAndTrailersForUnsupportedEncoding(
+    _ encoding: String,
+    advertisedEncoding: [String]
+  ) -> (GRPCStatus, acceptEncoding: String?) {
+    let status: GRPCStatus
+    let acceptEncoding: String?
+
+    if advertisedEncoding.isEmpty {
+      // No compression is supported; there's nothing to tell the client about.
+      status = GRPCStatus(code: .unimplemented, message: "compression is not supported")
+      acceptEncoding = nil
+    } else {
+      // Return a list of supported encodings which we advertise. (The list we advertise may be a
+      // subset of the encodings we support.)
+      acceptEncoding = advertisedEncoding.joined(separator: ",")
+      status = GRPCStatus(
+        code: .unimplemented,
+        message: "\(encoding) compression is not supported, supported algorithms are " +
+          "listed in '\(GRPCHeaderName.acceptEncoding)'"
+      )
+    }
+
+    return (status, acceptEncoding)
+  }
+
+  /// Extract and validate the 'content-type' sent by the client.
+  /// - Parameter headers: The headers to extract the 'content-type' from
+  private func extractContentType(from headers: HPACKHeaders) -> ContentType? {
+    return headers.first(name: GRPCHeaderName.contentType).flatMap(ContentType.init)
+  }
+
+  /// The result of validating the request encoding header.
+  private enum RequestEncodingValidation {
+    /// The encoding was valid.
+    case valid(messageReader: LengthPrefixedMessageReader, acceptEncoding: String?)
+    /// The encoding was invalid, the RPC should be terminated with this status.
+    case invalid(status: GRPCStatus, acceptEncoding: String?)
+  }
+
+  /// Extract and validate the request message encoding header.
+  /// - Parameters:
+  ///   - headers: The headers to extract the message encoding header from.
+  /// - Returns: `RequestEncodingValidation`, either a message reader suitable for decoding requests
+  ///   and an accept encoding response header if the request encoding was valid, or a pair of
+  ///     `GRPCStatus` and trailers to close the RPC with.
+  private func extractRequestEncoding(from headers: HPACKHeaders) -> RequestEncodingValidation {
+    let encodingHeader = headers.first(name: GRPCHeaderName.encoding)
```

Comment:
Is the use of `first` safe here? Are we choosing to ignore other appearances of this header on purpose?

Reply:
Good question. We don't expect other occurrences of this header which is why I used `first`. Do you think we should use `canonicalForm` instead and close if there's more than one value?

Reply:
I'd be inclined to do that, yeah.